invoke-ai/InvokeAI
# Add `TiledMultiDiffusionDenoiseLatents` invocation (for upscaling workflows)

## Summary

This PR adds a `TiledMultiDiffusionDenoiseLatents` invocation. This invocation has an interface similar to that of the `DenoiseLatentsInvocation`, but uses [MultiDiffusion](https://arxiv.org/abs/2302.08113) internally to perform tiled denoising. The primary intended use case for this node is in tiled upscaling workflows.

 This PR refactors (and tidies) the `StableDiffusionGeneratorPipeline` class to facilitate sharing most of the core logic between standard and MultiDiffusion denoising.

### Workflows

Here are some test workflows that use the new invocation:
- Tiled upscaling with ESRGAN x2 as the initial upscaler: 
[sd15_multi_diffusion_esrgan_x2_upscale.json](https://github.com/user-attachments/files/15975542/sd15_multi_diffusion_esrgan_x2_upscale.json)
- A simple panorama workflow for more direct testing of MultiDiffusion: [multi_diffusion_panorama.json](https://github.com/user-attachments/files/15905115/multi_diffusion_panorama.json)

Screenshot of multi_diffusion_esrgan_x2_upscaling.json:
<img width="1722" alt="image" src="https://github.com/invoke-ai/InvokeAI/assets/14897797/7d41a675-2ee0-4e09-96c4-f4d2335d9c72">

### Sample Outputs from Upscaling

Before (352x352):
![image](https://github.com/invoke-ai/InvokeAI/assets/14897797/fb2576f2-e807-436f-9774-b7bc3ced0e89)

After 4x (1408x1408):
![image](https://github.com/invoke-ai/InvokeAI/assets/14897797/64fca600-aecf-4658-8ee5-b95c6efe40d4)

Before (842x842):
![image](https://github.com/invoke-ai/InvokeAI/assets/14897797/ca669c30-5dee-4ea5-8a35-762b559887cd)

After 4x (3360x3360):
![3c5f04e9-c60a-4ee5-84ee-b624b19b3bab](https://github.com/invoke-ai/InvokeAI/assets/14897797/ad54fbbd-1c27-42a3-9891-bf56b323ec09)

## QA Instructions

Test plan:
- SD 1.5 regression tests (no change expected on these workflows with this PR)
  - [x] Text-to-image
  - [x] Image-to-image
  - [x] Inpainting with non-inpainting model
  - [x] Inpainting with inpainting model
  - [x] Text-to-image with an inpainting model
  - [x] Regional text conditioning and regional IP-Adapter (just a smoke test - this shouldn't be affected)
- SDXL regression tests (no change expected on these workflows with this PR)
  - [x] Text-to-image
  - [x] Image-to-image
  - [x] Inpainting with non-inpainting model
- Multi-Diffusion tests
  - [x] SD 1.5 upscaling workflow
  - [x] SDXL upscaling workflow
  - [x] Try to break it with awkward input image dimensions

## Merge Plan



## Checklist
